### ISSUE TITLE
Linux Support

### ISSUE THREAD

**Issue by reahari** [tooll3/t3#77]
Are there any plans to support linux?

**Comment by pixtur** [comment]
Hi @reahari  thanks for your question. I wrote my current thoughts here: https://github.com/still-scene/t3/wiki/faq#so-where-are-the-maclinux-versions

**Comment by NicTanghe** [comment]
Hello everyone, please send any developers that know their way around Vulcan this way.
I dont like using wine for audio using apps. and we all want this to run with vulcan.

**Comment by pixtur** [comment]
Thanks for looking into this. Linux support would indeed be great!

**Comment by domportera** [comment]
I'm recently trying to migrate over to linux, so my interest in this is piqued. fortunately, [DXVK](https://github.com/doitsujin/dxvk) exists now and is always improving thanks to the greater Wine and Proton communities. So I'd be interested in taking a stab at getting it working - it seems like it's a simple .dll replacement. If I can manage to wrangle the linking such that it's only used on non-Windows builds, it could be a simple solution backed by a huge community. 

The other consideration is, of course, UI. There's lots of options with various pros, cons, and licenses. I see there being three potential choices:

- [Uno](https://github.com/unoplatform/uno) - Apache 2.0 license
- [Avalonia](https://github.com/AvaloniaUI/Avalonia) - MIT license - probably the best, sans the hot-reloading UI Uno has
- [Eto](https://github.com/picoe/Eto) - BSD 3 license (similar to MIT) - uses WinForms for Windows builds, code structure is also very similar. But might be a little buggier than the others with less support.

I looked into plain ol GTK C# wrappers and they're pretty unmaintained, so I don't see that as a great option.

**Comment by NicTanghe** [comment]
Y close the ticket when there is no linux support ?
Just leave it open ?

**Comment by pixtur** [comment]
Ups. Closed this by mistake. Mixed up some doubles.

**Comment by m1karii** [comment]
@domportera +1 for Avalonia

**Comment by d-franklin** [comment]
Came across this project today, looks fantastic. Great work!

Another potential solution for Linux and mac support is using [Godot](https://godotengine.org) which is an open source game engine which supports C# (as well as C++ and their own native language).

While a game engine might seem like a weird solution, Godot already supports the following out of the box.

* Windows, Linux, mac, android and ios
* Volkan
* Shaders
* 2D and 3D rendering
* 2D and 3D animation
* UX components with the ability for complete customization of the theme
* Audio
* Video (although default video is limited, an extension that uses ffmpeg is available)
* As well as a host of other features

While it would mean a rewrite of sections of the code, I think it's worth investigating as a potential solution since both Godot and Tooll3 are powerful engines/tools which could complement each other.

**Comment by reahari** [comment]
@d-franklin I don't think that's a good solution as it would require an entire rewrite

**Comment by d-franklin** [comment]
> @d-franklin I don't think that's a good solution as it would require an entire rewrite

Any solution adding support for Linux/Mac will require a substantial rewrite.

**Comment by domportera** [comment]
it's funny you mention that @d-frankli - @pixtur and I have talked about exactly this before.

Currently I think the plan is to just continue chopping away at pieces of Tooll to allow for multiplatform support / embedded Tooll in the future. This will take some time but in the future I would imagine cross-platform support be more native a la Avalonia/Vulkan/WebGPU, but personally I'm also trying to work towards allowing tooll to be embedded within game engines like Godot and pixtur is excited for that possibility as well!

the general idea is to abstract out the graphics implementation to not be reliant on SharpDX as it is currently. Progress on this has started but it is a long road ahead 😩

**Comment by Merlin04** [comment]
Was able to get it running in wine pretty easily (just had to install some .net runtime stuff linked in the error messages it was giving), it seems like basic functions are working except video export (it just says it was unsuccessful). If Tooll could officially support running under wine it might be a lot easier than implementing full native linux support!

**Comment by d-franklin** [comment]
> Was able to get it running in wine pretty easily (just had to install some .net runtime stuff linked in the error messages it was giving), it seems like basic functions are working except video export (it just says it was unsuccessful). If Tooll could officially support running under wine it might be a lot easier than implementing full native linux support!

Yer this should probably be closed this since wine is a working solution for Linux.

**Comment by dromer** [comment]
Hmm, using wine is a pretty poor "solution" for this. Native support would most certainly be a lot better.

Of course it would require quite some work, but flat-out closing the issue like this is a bit sad.
Someone else could see the issue and be motivated to do the work. Now this visibility is just gone.

**Comment by Merlin04** [comment]
Yeah, I also don't think it should be closed, especially given that Wine isn't a working solution for Linux yet - video export doesn't work, which is a pretty important feature!

**Comment by domportera** [comment]
idk why this was closed, but as the dude kinda fronting the tooll refactor initiative atm, I can assure you that Linux is not off the todo list - native linux. that being said it's looking like it's gonna be a long time bc of the challenge and lack of libraries available for Vulkan in C# at the moment

as for the video export - it sounds like a MediaFoundation issue. Doing some Wine research around that might yield some results or info? I'd be curious to know if y'all find anything

**Comment by enderprism** [comment]
is the reason why video exporting on Wine doesn't work known? i started using Tooll today on my EndeavourOS machine, and followed the wiki page to install Tooll
i thought it could be related to ffmpeg, so i installed it in the bottle's wineprefix, and added it to its `PATH`, to no avail

**Comment by domportera** [comment]
not sure exactly how to get it going, but see if you could include Windows MediaFoundation in you wine bottle and pls let me know if that fixes it!

**Comment by enderprism** [comment]
i'll look into it asap

**Comment by enderprism** [comment]
i installed MediaFoundation through the bottle's dependencies, which didn't fix anything (and video importing is broken too but i assume it also relies on MediaFoundation)
i also tried to install MediaFoundation with [`muink/mf-install'`](https://github.com/muink/mf-install)s install script because it looked like it had more DLLs that maybe were missing from Bottles' install script, and it didn't work either.

![Image](https://github.com/user-attachments/assets/5ffc4571-ed86-4275-bf46-e650cd494abc)

![Image](https://github.com/user-attachments/assets/71acb312-b4c9-4467-a692-347d95648ac1)

**Comment by domportera** [comment]
damn, good to know - thank you for getting back with a definite answer! 

yes all video stuff is MediaFoundation at the moment unfortunately. swapping out the video framework is a pretty challenging thing to do, but we've been considering it in the medium term. the fact that adding that dependency to wine doesn't resolve the problem adds some urgency to the issue. 

it's curious though that the editor doesn't crash when rendering - it just doesn't do anything?

**Comment by domportera** [comment]
also, since the script you mentioned isn't necessary in games anymore, that gives me an idea - could you try running tooll in a proton bottle? even if it's just adding t3 to your steam library for a moment 😅

**Comment by enderprism** [comment]
sure thing, will do
iirc you can Bottles supports proton-ge as one of its runners so i'll use that instead of moving the prefix to steam

**Comment by enderprism** [comment]
changing the bottle's runner to proton-ge and starting t3 inside it fails and yields an error related to libicu.

![Image](https://github.com/user-attachments/assets/01461e1a-297d-4e45-b463-49d0945f4e5c)

**Comment by enderprism** [comment]
i don't think it's a proton issue however, i remember adding Tooll as an external program to steam and setting the runner to proton worked fine
let me test that again

**Comment by enderprism** [comment]
importing Tooll as a non-steam program, using protontricks to add `dotnetdesktop6`, `d3dcompiler_46`, `d3dcompiler_47`, `d3dcompiler_43` and `mf` to the prefix, and running t3 using Proton 9 does not fix anything.
(i installed all d3dcompiler versions available because some shaders failed to compile with only `d3dcompiler_46` — i guess Bottles does more than installing this single version)

**Comment by domportera** [comment]
thank you for clarifying! then this is absolutely something that won't be fixed by proton any time soon, and if it is then it won't exactly be user-friendly anyway. changing the video framework seems like the only way

**Comment by enderprism** [comment]
glad i could be of any help! it's kind of unfortunate we won't have video exporting until that happens though

**Comment by enderprism** [comment]
we can also render as an image sequence and edit the audio in with other software!

i wasn't sure there would be an option to render as an image sequence because its window is separate to the video render, for some reason.
they could be merged into a single window with a `Render as` dropdown option to change between `Image Sequence` or `Video` (just my grain of salt) so that this render type is exposed to new users without needing to search in menus.

**Comment by enderprism** [comment]
i wanted to give 4.0.0 a try (using the build from the releases page), but it can't seem to create any projects, and the project list is empty.
i remember it showed some kind of error when installing tixl but i can't remember what it said, it went by too fast.

![Image](https://github.com/user-attachments/assets/40cd5717-21b8-47a4-bf90-73569240937b)

![Image](https://github.com/user-attachments/assets/b83d6441-c8db-4732-b953-e511d367a044)

<details><summary>
<code>winetricks list-installed</code> output</summary>
<p>

```
Using winetricks 20250102 - sha256sum: 53194dead910f8a5eb1deacaa4773d4e48f5873633d18ab1ecd6fdb0cb92243b with wine-10.6 and WINEARCH=win64
dotnet6
dotnetdesktop6
d3dcompiler_46
d3dcompiler_43
d3dcompiler_47
dotnetdesktop9
powershell_core
powershell
```

</p>
</details>

**Comment by pixtur** [comment]
This is might have been an issue with v4.0.0.
Could you try with v4.0.1.1

**Comment by enderprism** [comment]
i'll try it asap

**Comment by enderprism** [comment]
it works flawlessly!

**Comment by wargreen** [comment]
The 
```
Failed to create project "T1" in "".

This should never happen - please file a bug report.
Currently this error is unhandled, so you will want to manually delete the project from disk.
```
error seem to be back in 4.0.2

 Additional info (if useful) 
```
Failed to start cross-platform powershell: An error occurred trying to start process 'pwsh' with working directory 'C:\users\wargreen\AppData\Roaming\TiXL\Tmp\CompilationWorkingDirectory'. Fichier introuvable.
```

EDIT : I've installed 4.0.1.1 and i got the same error

**Comment by domportera** [comment]
possibly a dumb question, but do you have powershell on your system?

**Comment by wargreen** [comment]
> possibly a dumb question, but do you have powershell on your system?

Not so dumb... Installing powershell in the prefix with winetricks fix the issue.
I thought that powershell come with .net.

I put all the info for run it in wine in my wiki : https://l2m2p.lebib.org/?Tooll32

**Comment by domportera** [comment]
powershell is based on/interoperable with .net but nah it's a whole other external application/shell

**Comment by anpin** [comment]
Would love to see native linux support. 

@domportera have you considered to port SharpDX bindings to Silk.NET? There is a [WIP PR in the Stride repo](https://github.com/stride3d/stride/pull/1715) where they try to attempt just that.

**Comment by domportera** [comment]
> Would love to see native linux support. 
> 
> @domportera have you considered to port SharpDX bindings to Silk.NET? There is a [WIP PR in the Stride repo](https://github.com/stride3d/stride/pull/1715) where they try to attempt just that. 

thats my personal preference yes! long term would look like (in order-ish):

- switch to Silk DirectX
- switch to Silk Windowing and Input
- migrate to cross platform video API
- switch to silk vulkan (maybe using something like Daxa as an inbetween)

that being said id rather wait until Silk 3.0 to commit to it - it's currently waiting on me to finish the input implementation 🙈

**Comment by enderprism** [comment]
how hard would it be to rewrite RenderExport to use something like ffmpeg instead of mediafoundation? it's [currently impossible](https://github.com/tixl3d/tixl/issues/909#issuecomment-3725101134) to render out videos due to a half-implemented MF function in WINE (`SinkWriter.SetInputMediaType`). maybe switching to something more system-agnostic could help.
